neqo-transport/src/path.rs

```diff
@@ -749,15 +764,25 @@ impl Path {
 
     /// At the next opportunity, send a probe.
     /// If the probe count has been exhausted already, marks the path as failed.
-    fn probe(&mut self) {
+    fn probe(&mut self, stats: &mut Stats) {
         let probe_count = match &self.state {
             ProbeState::Probing { probe_count, .. } => *probe_count + 1,
             ProbeState::ProbeNeeded { probe_count, .. } => *probe_count,
             _ => 0,
         };
         self.state = if probe_count >= MAX_PATH_PROBES {
-            qinfo!([self], "Probing failed");
-            ProbeState::Failed
+            if self.ecn_info.ecn_mark() == IpTosEcn::Ect0 {
+                // The path validation failure may be due to ECN blackholing, try again without ECN.
+                qinfo!(
+                    [self],
+                    "Possible ECN blackhole, disabling ECN and re-probing path"
+                );
+                self.ecn_info.disable_ecn(stats);
+                ProbeState::ProbeNeeded { probe_count: 0 }
+            } else {
+                qinfo!([self], "Probing failed");
+                ProbeState::Failed
+            }
```

Comment:
Question for my own understanding:

- Ideally we would bundle all ECN related logic in `neqo_transport::ecn`.
- `neqo_transport::ecn` already has the notion of failing ECN validation early, i.e. on connection establishment. See `ECN_TEST_COUNT_INITIAL_PHASE` introduced in https://github.com/mozilla/neqo/pull/1964.
- It would be nice if we wouldn't have two cases of early ECN validation failure logic, (1) for connection establishment and (2) for path validation. But instead for both to be handled by the same logic.
- (1) is triggered by a loss-recovery timeout, determining the ECN marked packets as lost (see `EcnInfo::on_packets_lost`).
- (2) can not be triggered by loss-recovery, given that there is no timeout mechanism for non-primary paths, thus probe packets are never determined as lost and thus `EcnInfo::on_packets_lost` is never called.

Is that correct @larseggert?

Reply:
Yes. I agree that a consolidated approach would be nicer.